src/conv-core/isomalloc.C

```diff
@@ -896,19 +896,210 @@ static void CmiIsomallocInitExtent()
+  {
+    IMP_DBG("[%d][%p] isommap::isommap(%p, %p)\n", CmiMyPe(), this, s, e);
+  }
+  isommap(PUP::reconstruct pr)
+    : lock{CmiCreateLock()}
+  {
+    IMP_DBG("[%d][%p] isommap::isommap(PUP::reconstruct)\n", CmiMyPe(), this);
+  }
+  ~isommap()
+  {
+    IMP_DBG("[%d][%p] isommap::~isommap()\n", CmiMyPe(), this);
+    clear();
+    CmiDestroyLock(lock);
+  }
+
+  void print() const
+  {
+    CmiPrintf("[%d][%p] isommap::print(): %p-%p %zu\n", CmiMyPe(), this, start, allocated_extent, allocated_extent - start);
+  }
+
+  bool isInRange(void * user_ptr) const
+  {
+    return start <= (const uint8_t *)user_ptr && (const uint8_t *)user_ptr < end;
+  }
+  bool isMapped(void * user_ptr) const
+  {
+    return start <= (const uint8_t *)user_ptr && (const uint8_t *)user_ptr < allocated_extent;
+  }
+
+  void pup(PUP::er & p)
+  {
+    IMP_DBG("[%d][%p] isommap::pup()%s%s%s%s\n", CmiMyPe(), this, p.isSizing() ? " sizing" : "",
+            p.isPacking() ? " packing" : "", p.isUnpacking() ? " unpacking" : "", p.isDeleting() ? " deleting" : "");
+
+    pup_raw_pointer(p, start);
+    pup_raw_pointer(p, end);
+    pup_raw_pointer(p, allocated_extent);
+
+    const size_t totalsize = allocated_extent - start;
+
+    if (p.isUnpacking())
+    {
+      void * const mapped = map_global_memory(start, totalsize);
+      if (mapped == nullptr)
+        CmiAbort("Failed to unpack Isomalloc memory region!");
+    }
+
+    p(start, totalsize);
+
+    if (p.isDeleting())
+    {
+      clear();
+    }
+  }
+
+  void clear()
+  {
+    if (allocated_extent == start)
+      return;
+    unmap_global_memory(start, allocated_extent - start);
+    allocated_extent = start;
+  }
+
+  // canonical data
+  uint8_t * start, * end, * allocated_extent;
+
+  // local data
+  CmiNodeLock lock;
+};
+
+/************** dlmalloc mempool ***************/
+
+#include "memory-gnu-internal.C"
+
+struct isomalloc_dlmalloc : dlmalloc_impl
+{
+  isomalloc_dlmalloc(uint8_t * s, uint8_t * e)
+    : backend{s, e}, arena{create_mspace(0, 0)}
+  {
+    IMP_DBG("[%d][%p] isomalloc_dlmalloc::isomalloc_dlmalloc(%p, %p)\n", CmiMyPe(), this, s, e);
+  }
+  isomalloc_dlmalloc(PUP::reconstruct pr)
+    : backend{pr}
+  {
+    IMP_DBG("[%d][%p] isomalloc_dlmalloc::isomalloc_dlmalloc(PUP::reconstruct)\n", CmiMyPe(), this);
+  }
+
+  void pup(PUP::er & p)
+  {
+    p | backend;
+    pup_raw_pointer(p, arena);
+  }
+
+  isommap backend;
+  mspace arena;
+
+  virtual void * call_mmap(size_t length) override final
+  {
+    void * const mapped = map_global_memory(backend.allocated_extent, length);
+    if (mapped == nullptr)
+      return nullptr;
+    backend.allocated_extent += length;
+    return mapped;
+  }
+  virtual void * call_direct_mmap(size_t length) override final
+  {
+    return this->isomalloc_dlmalloc::call_mmap(length); // call call_mmap without the vtable lookup
+  }
+  virtual int call_munmap(void * addr, size_t length) override final
+  {
+    // TODO
```

Comment:
Is these something more to do here?

Reply:
This implementation currently meets or exceeds what the Isomempool does. The `TODO` refers to  maintaining a record of mmapped regions so that migration can skip the gaps. Complexity arises when considering RDMA transfers, and we would need to discuss when it is better to invoke multiple transfers, or to leave small gaps mapped so they could be combined into one.

Reply:
I see, can you just put this note into the TODO comment?

Reply:
Done